packages/record-data/addon/-private/ordered-set.ts

```diff
@@ -1,22 +1,118 @@
 import { assert } from '@ember/debug';
-import { guidFor } from '@ember/object/internals';
-import EmberOrderedSet from '@ember/ordered-set';
 
-export default class EmberDataOrderedSet<T> extends EmberOrderedSet<T> {
-  static create() {
-    return new this();
+type Dict<T> = import('@ember-data/store/-private/ts-interfaces/utils').Dict<T>;
+type RelationshipRecordData = import('./ts-interfaces/relationship-record-data').RelationshipRecordData;
+
+const NULL_POINTER = `null-${Date.now()}`;
+
+/*
+ TODO Ember's guidFor returns a new string per-object reference
+ while ours does not.
+
+ This has surfaced a bug during resurrection
+ in which Ember's guidFor would return false for `has` since the
+ resurrected record receives a fresh RecordData instance, leaving
+ the destroyed record in the set and thus depending on the state flags
+ for it not appearing elsewhere. We've accounted for this bug in the
+ updated OrderedSet implementation by doing a reference check: e.g.
+ the bug is preserved.
+
+ While we convert relationships to identifiers this will be something we
+ will be forced to address.
+*/
+function guidFor(obj): string {
+  if (obj === null) {
+    return NULL_POINTER;
+  }
+
+  return obj.clientId || obj.lid;
+}
+
+export default class OrderedSet {
+  declare presenceSet: Dict<RelationshipRecordData | null>;
```

Comment:
`null` was a mistake in our typing, though we probably shouldn't have been allowing it to be put into the set. I think it happens mostly because we use these for `BelongsTo` relationships as well in which the canonical / current state can often be `null`.

The referenced bug is that we end up keeping destroyed recordData instances in the relationships because the check to remove them fails. This is because when we resurrect a record right before it's discarded we actually hand it a new RecordData instance. Whether the bug is RecordData being a new instance or the bug is not removing the old RecordData, it happens because Ember's `guidFor` builds a new key for each thing it see by-reference. We don't actually care about by-reference here, we care about identity. But for now we've had to preserve the by-reference behavior.